compiler/src/main/java/com/squareup/anvil/compiler/ModuleMerger.kt

```diff
@@ -92,112 +84,107 @@ internal class ModuleMerger(
         // issue we filter all generated Anvil modules except for the one that was generated for
         // this specific class.
-        val fqName = it.fqNameSafe
+        val fqName = it.fqName
         !fqName.isAnvilModule() || fqName == anvilModuleName
       }
-      .mapNotNull {
-        val contributesAnnotation =
-          it.annotationOrNull(contributesToFqName, scope = scopeFqName)
-            ?: return@mapNotNull null
-        it to contributesAnnotation
+      .mapNotNull { contributedClass ->
+        val contributesAnnotation = contributedClass.annotationOrNull(
+          annotationName = contributesToFqName,
+          scopeName = scopeFqName
+        ) ?: return@mapNotNull null
+
+        contributedClass to contributesAnnotation
       }
-      .filter { (classDescriptor, _) ->
-        val moduleAnnotation = classDescriptor.annotationOrNull(daggerModuleFqName)
-        val mergeModulesAnnotation = classDescriptor.annotationOrNull(mergeModulesFqName)
-        if (!DescriptorUtils.isInterface(classDescriptor) &&
+      .filter { (contributedClass, _) ->
+        val moduleAnnotation = contributedClass.annotationOrNull(daggerModuleFqName)
+        val mergeModulesAnnotation = contributedClass.annotationOrNull(mergeModulesFqName)
+        if (!contributedClass.isInterface() &&
           moduleAnnotation == null &&
           mergeModulesAnnotation == null
         ) {
-          throw AnvilCompilationException(
-            classDescriptor,
-            "${classDescriptor.fqNameSafe} is annotated with " +
+          throw AnvilCompilationExceptionClassReference(
+            contributedClass,
+            "${contributedClass.fqName} is annotated with " +
               "@${ContributesTo::class.simpleName}, but this class is neither an interface " +
               "nor a Dagger module. Did you forget to add @${Module::class.simpleName}?"
           )
         }
 
         moduleAnnotation != null || mergeModulesAnnotation != null
       }
-      .onEach { (classDescriptor, _) ->
-        if (classDescriptor.effectiveVisibility() !is Public) {
-          throw AnvilCompilationException(
-            classDescriptor,
-            "${classDescriptor.fqNameSafe} is contributed to the Dagger graph, but the " +
-              "module is not public. Only public modules are supported."
-          )
+      .onEach { (contributedClass, _) ->
+        contributedClass.checkClassIsPublic {
+          "${contributedClass.fqName} is contributed to the Dagger graph, but the " +
+            "module is not public. Only public modules are supported."
         }
       }
       // Convert the sequence to a list to avoid iterating it twice. We use the result twice
       // for replaced classes and the final result.
       .toList()
 
-    val excludedModules = (mergeAnnotation.getAnnotationValue("exclude") as? ArrayValue)
+    val excludedModules = (mergeAnnotation.annotation.getAnnotationValue("exclude") as? ArrayValue)
```

Comment:
`AnnotationReference` gives you the excluded values. No need to work with descriptors.